docs/lang/logic.md

```diff
@@ -0,0 +1,458 @@
+
+**Effect:** This operator is a magical device by the definition, no kidding. It
+is a blackbox algorithm that _somehow_ picks one element `e` of `S` that
+satisfies the expression `P` against the binding `[x |-> e]` and returns this
+element `e`.
+
+Did we say "an algorithm"? Yes! `CHOOSE x \in S: P` is deterministic.  If you
+give it two equal sets and two equivalent predicates, it will give you the
+same value. More formally, the only known property of `CHOOSE` is as follows:
+
+```tla
+  {x \in S: P} = {y \in T: Q} =>
+      (CHOOSE x \in S: P) = (CHOOSE y \in T: Q)
+```
+
+Importantly, when `{x \in S: P} = {}`, the expression `CHOOSE x \ in S: P`
+    evaluates to an undefined value.
+
+How does it work? We don't know. Nobody knows. Maybe it returns the first
+element of the set? Well, sets are not ordered, so there is no first element.
+
+Why shall you use it? Actually, you should not. Unless you have no other
+choice :bowtie:
+
+There are two common use cases, where the use of `CHOOSE` is well justified:
+
+ - _Use case 1:_ Retrieving the only element of a singleton set. If you know
+    that `Cardinality({x \in S: P}) = 1`, then `CHOOSE x \in S: P` returns
+    the only element of `{x \in S: P}`. No magic.
+    For instance, see: [Min and Max in FiniteSetsExt].
+
+ - _Use case 2:_ Enumerating set elements in a fixed but unknown order.
+    For instance, see: [ReduceSet in FiniteSetsExt].
+
+In other cases, we believe that `CHOOSE` is bound to do [Program synthesis].
+So TLC does some form of synthesis by brute force when it has to evaluate
+`CHOOSE`.
+   
+**Determinism:** Deterministic. Very much deterministic. Don't try to model
+non-determinism with `CHOOSE`. For non-determinism, see:
+[Control Flow and Non-determinism].
+
+Apalache picks a set element that satisfies the predicate `P`, but it does not
+guarantee the repeatability property of CHOOSE. It does not guarantee
+non-determinism either. Interestingly, this behavior does not really make a
+difference for the use cases 1 and 2. If you believe that this causes a problem
+in your specification, [open an issue...]
+
+**Errors:** Pure TLA+ does not restrict the operator arguments.  TLC flags a
+model checking error, if `S` is infinite.  Apalache produces a static type
+error, if the type of elements of `S` is not compatible in the context of `P`
+when an element of `S` is bound to `x`.
+
+**Example in TLA+:**
+
+```tla
+  CHOOSE x \in 1..3: x >= 3
+  \* 3
+  CHOOSE x \in 1..3:
+    \A y \in 1..3: y >= x
+  \* 1, the minimum
+  CHOOSE f \in [ 1..10 -> BOOLEAN ]:
+    \E x, y \in DOMAIN f:
+      f[x] /\ ~f[y]
+  \* some Boolean function from 1..10 that covers FALSE and TRUE
+```
+
+**Example in Python:** Python does not have anything similar to `CHOOSE`.
+The closest possible solution is to sort the filtered set by the string values
+and pick the first one (or the last one). So we have introduced a particular
+way of implementing CHOOSE, see [choose.py](./examples/choose.py):
```

Comment:
You can do:
```python
def get(s):
  for e in s:
    return e

print(get({"1", 1, (1,)}))
```
under the hood, set is iterable, so you get the 1st element of that iteration.

Reply:
Yes, but that would not guarantee the determinism of CHOOSE. I am enforcing determinism by sorting the elements with their string representation as a key. It is quite inefficient, but sound.